### PR TITLE
Add support for Stringified NBT objects

### PR DESCRIPTION
**Pull request by aakatz3**
Cool new system to make building NBT strings easy and not awful

### CODE REVIEW

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtBooleanTag.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import com.google.common.base.Preconditions;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonPrimitive;
+import org.jetbrains.annotations.NotNull;
+
+
+/**
+ * A convenience class for converting booleans from {@link JsonPrimitive}s, and for easily creating
+ */
+public class NbtBooleanTag extends NbtPrimitiveTag<Boolean> {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtBooleanTag.java:15]:
```suggestion
public class NBTBooleanTag extends NBTPrimitiveTag<Boolean> {
```
This is personal preference I guess, but having abbreviations in class names be uppercase appears cleaner imo

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtBooleanTag.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        return bool;
+    }
+
+    @Override
+    public @NotNull JsonElement toJson() {
+        return new JsonPrimitive(bool);
+    }
+
+    @Override
+    public @NotNull String toSNbt() {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtBooleanTag.java]:
```suggestion
    public @NotNull String toString() {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtCompoundTag.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.StringJoiner;
+
+/**
+ * A wrapper tag that allows working with Compound NBT tags as maps!
+ */
+public class NbtCompoundTag extends HashMap<String, NbtTag> implements NbtTag{
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtCompoundTag.java]:
```suggestion
public class NBTCompoundTag extends HashMap<String, NBTTag> implements NBTTag {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtCompoundTag.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                }
+            } else {
+                object.add(name, nbtTag.toJson());
+            }
+        });
+        return object;
+    }
+
+    @Override
+    public @NotNull String toSNbt() {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtCompoundTag.java]:
```suggestion
    public @NotNull String toString() {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonAdapter.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import lombok.experimental.UtilityClass;
+import org.jetbrains.annotations.NotNull;
+
+import java.text.MessageFormat;
+
+/**
+ * Utility class for easy creation of the appropriate {@link NbtTag} from an existing {@link JsonElement}
+ */
+@UtilityClass
+public final class NbtJsonAdapter {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonAdapter.java:18]:
```suggestion
public final class NBTJsonAdapter {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonAdapter.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    /**
+     * Creates the appropriate tag from a given {@link JsonElement}, but never creates a {@link NbtJsonTag}
+     * @param jse The incoming {@link JsonElement} to convert
+     * @return the new {@link NbtTag}
+     */
+    public static NbtTag jsonToNbt(@NotNull JsonElement jse) {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonAdapter.java]:
```suggestion
    public static NbtTag fromJson(@NotNull JsonElement element) {
```
again personal preference, to make it look cleaner

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonAdapter.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        } else if(jse.isJsonPrimitive()){
+            JsonPrimitive primitive = jse.getAsJsonPrimitive();
+            if(primitive.isBoolean()){
+                return NbtBooleanTag.fromPrimitive(primitive);
+            } else if(primitive.isNumber()){
+                return NbtNumberTag.fromPrimitive(primitive);
+            } else if(primitive.isString()){
+                return NbtStringTag.fromPrimitive(primitive);
+            } else {
+                throw new UnsupportedOperationException("Illegal type of NBT primitive");
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonAdapter.java]: if possible, please use a switch statement here, to make it more readable :)

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonTag.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonElement;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A special type of {@link NbtTag} that converts an arbitrary {@link JsonElement} into an appropriately escaped string as SNBT
+ */
+public class NbtJsonTag implements NbtTag{
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonTag.java:15]:
```suggestion
public class NBTJsonTag implements NBTTag {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonTag.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    @Override
+    public @NotNull JsonElement toJson() {
+        return element;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public @NotNull String toSNbt() {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonTag.java]:
```suggestion
    public @NotNull String toString() {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtListTag.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.StringJoiner;
+
+/**
+ * A wrapper class that allows the construction of or conversion (from JSON) of SNBT List Tags
+ */
+
+public class NbtListTag extends ArrayList<NbtTag> implements NbtTag  {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtListTag.java:19]:
```suggestion
public class NBTListTag extends ArrayList<NBTTag> implements NBTTag {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtListTag.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    public @NotNull JsonElement toJson() {
+        JsonArray array = new JsonArray();
+        for(NbtTag tag : this){
+            array.add(tag.toJson());
+        }
+        return array;
+    }
+
+    @Override
+    public @NotNull String toSNbt() {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtListTag.java]:
```suggestion
    public @NotNull String toString() {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtNumberTag.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+import java.text.DecimalFormat;
+import java.text.DecimalFormatSymbols;
+import java.text.MessageFormat;
+import java.text.NumberFormat;
+import java.util.Locale;
+
+/**
+ * A wrapper class that allows you to deal with numbers in NBT
+ */
+public class NbtNumberTag extends NbtPrimitiveTag<Number> {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtNumberTag.java:20]:
```suggestion
public class NBTNumberTag extends NBTPrimitiveTag<Number> {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtNumberTag.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    @Override
+    public @NotNull JsonElement toJson() {
+        return new JsonPrimitive(number);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public @NotNull String toSNbt() {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtNumberTag.java]:
```suggestion
    public @NotNull String toString() {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtPrimitiveTag.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+package com.dumbdogdiner.stickyapi.common.nbt;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Base class for any NBT tags that wrap simple primitives (and strings)
+ *
+ * @param <T> The type of primitive
+ */
+public abstract class NbtPrimitiveTag<T> implements NbtTag {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtPrimitiveTag.java:14]:
```suggestion
public abstract class NBTPrimitiveTag<T> implements NBTTag {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtStringTag.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+import com.dumbdogdiner.stickyapi.common.util.StringUtil;
+import com.google.common.base.Preconditions;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonPrimitive;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A {@link NbtTag} that wraps and escapes {@link String}s
+ */
+public class NbtStringTag extends NbtPrimitiveTag<String> {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtStringTag.java:16]:
```suggestion
public class NBTStringTag extends NBTPrimitiveTag<String> {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtStringTag.java [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    @Override
+    public @NotNull JsonElement toJson() {
+        return new JsonPrimitive(string);
+    }
+
+    @SuppressWarnings("UnnecessaryStringEscape")
+    @Override
+    public @NotNull String toSNbt() {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtStringTag.java]:
```suggestion
    public @NotNull String toString() {
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtStringTag.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+     * @throws IllegalArgumentException if the primitive type is wrong
+     * @see JsonPrimitive#isString()
+     */
+    public static NbtStringTag fromPrimitive(JsonPrimitive primitive) throws IllegalArgumentException {
+        Preconditions.checkArgument(primitive.isString(), "The primitive must be a string");
+        return new NbtStringTag(primitive.getAsString());
+    }
+
+    @Override
+    public String toString() {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtStringTag.java:53]:
```suggestion
    public String getRawString() {
```
obviously this conflicts with my suggestion to change `toSNBT()` to `toString()`, so this should be the raw string?

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtTag.java [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+     */
+    @NotNull
+    JsonElement toJson();
+
+    /**
+     * Converts a tag into <a href="https://minecraft.gamepedia.com/NBT_format">Stringified NBT</a>
+     * @return A representation of the object as Stringified NBT (SNbt)
+     */
+    @NotNull
+    String toSNbt();
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtTag.java]:
```suggestion
    String toString();
```

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtTag.java [hunk 19]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2020-2021 DumbDogDiner <dumbdogdiner.com>. All rights reserved.
+ * Licensed under the MIT license, see LICENSE for more information...
+ */
+package com.dumbdogdiner.stickyapi.common.nbt;
+
+import com.google.gson.JsonElement;
+import org.jetbrains.annotations.NotNull;
+
+public interface NbtTag {
```
Comment (kokumaji) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtTag.java:10]:
```suggestion
public interface NBTTag {
```

**Review by kokumaji — DISMISSED** [review]
PR looks fine to me, except for the class names and some method names, would suggest reading up on Java's naming conventions (or we should agree on naming rules in the next meeting?). 

Waiting on other admins to comment on this PR.

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonAdapter.java (hunk 7 above)
Comment (aakatz3) [reply src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonAdapter.java]: I really wish I could, but sadly I cannot think of a way to. You sorta can with kotlin I think, but I'm not fully sure

**Comment by aakatz3** [comment]
I used `toSNBT` instead of `toString` to be explicit that it's converting to stringifies NBT/mojangson, rather than a JSON string or something else. I can change if people thing it would be better

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonAdapter.java (hunk 7 above)
Comment (kokumaji) [reply src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtJsonAdapter.java]: then at least collapse it to something like this? There doesn't seem to be a `getType()` method sadly, so this is the only thing we can do to make it appear cleaner
```java
        if(jse.isJsonArray()){
            return NbtListTag.fromJsonArray(jse.getAsJsonArray());
        } else if(jse.isJsonObject()){
            return NbtCompoundTag.fromJsonObject(jse.getAsJsonObject());
        } else if(jse.isJsonPrimitive()){
            JsonPrimitive primitive = jse.getAsJsonPrimitive();
            if(primitive.isBoolean() || primitive.isNumber() || primitive.isString()) {
                return NbtBooleanTag.fromPrimitive(primitive);
            else {
                throw new UnsupportedOperationException("Illegal type of NBT primitive");
```

**Comment by jcxldn** [comment]
> I used `toSNBT` instead of `toString` to be explicit that it's converting to stringifies NBT/mojangson, rather than a JSON string or something else. I can change if people thing it would be better

I wonder if something like `toNBTString()` would be a better name, should be the best of both worlds?

**Comment by aakatz3** [comment]
Sure! we good for StringNBT then?

**Comment by jcxldn** [comment]
`NbtBooleanTag` is correct for Google's Java Style Guide which we will be looking to implement, so should we go with that instead of `NBTBooleanTag`?

src: https://google.github.io/styleguide/javaguide.html#s5.2.2-class-names (UpperCamelCase class naming)

**Comment by kaylendog** [comment]
Merging on hold until after module refactor.

File: src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtCompoundTag.java [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.StringJoiner;
+
+/**
+ * A wrapper tag that allows working with Compound NBT tags as maps!
+ */
+public class NbtCompoundTag extends HashMap<String, NbtTag> implements NbtTag{
```
Comment (Prouser123) [inline src/main/java/com/dumbdogdiner/stickyapi/common/nbt/NbtCompoundTag.java:19]: spacing

**Review by jcxldn — COMMENTED** [review]
Reviewed all files.

**Review by jcxldn — APPROVED** [review]
Looks good! (~~will look at github codeql next.~~ **done**)

**Comment by jcxldn** [comment]
# CodeQL

- [x] Inconsistent equals and hashCode (NbtNumberTag.java)
https://github.com/DumbDogDiner/StickyAPI/security/code-scanning/67?query=ref%3Arefs%2Fheads%2Ffeat%2Fstring-nbt-helpers
- [x] Overloaded equals (NbtBooleanTagTest.java)
https://github.com/DumbDogDiner/StickyAPI/security/code-scanning/61?query=ref%3Arefs%2Fheads%2Ffeat%2Fstring-nbt-helpers
- [x] Inconsistent equals and hashCode (NbtStringTag.java)
https://github.com/DumbDogDiner/StickyAPI/security/code-scanning/60?query=ref%3Arefs%2Fheads%2Ffeat%2Fstring-nbt-helpers
- [x] Inconsistent equals and hashCode (NbtJsonTag.java)
https://github.com/DumbDogDiner/StickyAPI/security/code-scanning/59?query=ref%3Arefs%2Fheads%2Ffeat%2Fstring-nbt-helpers
- [x] Inconsistent equals and hashCode (NbtBooleanTag.java)
https://github.com/DumbDogDiner/StickyAPI/security/code-scanning/58?query=ref%3Arefs%2Fheads%2Ffeat%2Fstring-nbt-helpers

**Review by jcxldn — APPROVED** [review]

**Pull request merged**